[save/load] save/load the optimizer parameters

kparichay opened the pull request:
save and load the optimizer parameters as well for continued training
Add additive option in neural to continue training from previous training
    
Resolves #172
    
Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

dongju-chae commented on nntrainer/src/optimizer.cpp:
```diff
@@ -189,4 +189,22 @@ int Optimizer::setProperty(std::vector<std::string> values) {
 }
+
+void Optimizer::read(std::ifstream &file) {
+  if (type == OptType::adam) {
```
If optimizer parameters can be different depending on `OptType`, how about saving/reading opt type as well, at the first byte?

Also, in case of old models without opt parameters, I think you can handle them like that `continue_train` is unsupported, by checking file positions.

kparichay replied:
I dont think I understand completely about support existing models?
But I agree with you to save the `OptType`. This will also help in verifying if the current model being loaded matches with the optimizer type.

dongju-chae replied:
Ah, I also don't know you need to support old models too. It's just my guess about CI's failures.

kparichay replied:
The CI's failures was due to existing bugs exposed because of this PR.
They have now been resolved with the new push.

dongju-chae replied:
Great!

kparichay commented:
V2:
setOptimizer() bug fix to be called with set=true for only fc layer and not other layers
now setOptimizer() for fc_layer is unique compared to virtual defined by its parent
added continued training property for the optimizer
added getSize in tensor
save optimizer type to verify that the loaded optimizer values can be used sensibly
if not loading, move the file seek ahead

kparichay commented:
rebased
